tests/sentry/api/test_paginator.py

```diff
@@ -91,31 +107,133 @@ def test_simple(self):
+        assert result3[0] == res2
+        assert result3.next
+        assert result3.prev
+
+        result4 = paginator.get_result(limit=1, cursor=result3.prev)
+        assert len(result4) == 1, result4
+        assert result4[0] == res1
+        assert result4.next
+        assert not result4.prev
+
+    def test_descending(self):
+        joined = timezone.now()
+
+        res1 = self.create_user('foo@example.com', date_joined=joined)
+        res2 = self.create_user('bar@example.com', date_joined=joined + timedelta(seconds=1))
+        res3 = self.create_user('baz@example.com', date_joined=joined + timedelta(seconds=2))
+
+        queryset = User.objects.all()
+
+        paginator = DateTimePaginator(queryset, '-date_joined')
         result1 = paginator.get_result(limit=1, cursor=None)
         assert len(result1) == 1, result1
-        assert result1[0] == res1
+        assert result1[0] == res3
         assert result1.next
         assert not result1.prev
 
         result2 = paginator.get_result(limit=2, cursor=result1.next)
         assert len(result2) == 2, result2
         assert result2[0] == res2
-        assert result2[1] == res3
+        assert result2[1] == res1
         assert not result2.next
         assert result2.prev
 
-        # this is not yet correct
         result3 = paginator.get_result(limit=2, cursor=result2.prev)
-        assert len(result3) == 1, list(result3)
-        assert result3[0] == res1
+        assert len(result3) == 1, result3
+        assert result3[0] == res3
         assert result3.next
         assert not result3.prev
+
+    def test_prev_descending_with_new(self):
+        joined = timezone.now()
+
+        res1 = self.create_user('foo@example.com', date_joined=joined)
+        res2 = self.create_user('bar@example.com', date_joined=joined + timedelta(seconds=1))
+
+        queryset = User.objects.all()
+
+        paginator = DateTimePaginator(queryset, '-date_joined')
+        result1 = paginator.get_result(limit=10, cursor=None)
+        assert len(result1) == 2, result1
+        assert result1[0] == res2
+        assert result1[1] == res1
+
+        res3 = self.create_user('baz@example.com', date_joined=joined + timedelta(seconds=2))
+        res4 = self.create_user('qux@example.com', date_joined=joined + timedelta(seconds=3))
+
+        result2 = paginator.get_result(limit=10, cursor=result1.prev)
+        assert len(result2) == 2, result2
+        assert result2[0] == res4
+        assert result2[1] == res3
+
+        result3 = paginator.get_result(limit=10, cursor=result2.prev)
+        assert len(result3) == 0, result3
+
+        result4 = paginator.get_result(limit=10, cursor=result1.next)
+        assert len(result4) == 0, result4
+
+    def test_roudning_offset(self):
+        joined = timezone.now()
+
+        res1 = self.create_user('foo@example.com', date_joined=joined)
+        res2 = self.create_user('bar@example.com', date_joined=joined + timedelta(microseconds=1))
```

Comment:
Yeah, why you doing just microseconds? MySQL doesn't support microsecond precision, so you'll lose order.

Reply:
[MySQL 5.7 does seem to support millisecond accuracy](https://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html), but that's not good enough.

I think I'll x mark this test. If possible I'd like to only x mark it if we're running the MySQL DB tests.

Reply:
Skipped, actually, not x failed.

Reply:
Travis is definitely not running 5.7. 5.6 I think.

Either way, I guess I'm asking, what's the point of the test? Is there something explicitly you're trying to test with microsecond precision?

Reply:
Yes, because the cursors value is floored / ceild, the cursor builder has logic to add offsets to  skip values that fall within that precision gap so that when querying for the next result set, we ensure that we skip results we've already seen. This tests that we actually skip those results.